File: Leetcode/src/com/leetcode/graph/_399_EvaluateDivision.java
package com.leetcode.graph;

import java.util.*;

/**
 * @Date: 08/23/2020
 * @Description: Graph, DFS
 **/
public class _399_EvaluateDivision {
    // time:O(V + E) space:O(n)
    public double[] calcEquation(List<List<String>> equations, double[] values, List<List<String>> queries) {
        HashMap<String, HashMap<String, Double>> map = new HashMap<>();
        // build the graph
        for (int i = 0; i < equations.size(); i++) {
            String x = equations.get(i).get(0);
            String y = equations.get(i).get(1);
            double val = values[i];
            if (!map.containsKey(x)) map.put(x, new HashMap<String, Double>());
            map.get(x).put(y, val);
            if (!map.containsKey(y)) map.put(y, new HashMap<String, Double>());
            map.get(y).put(x, 1 / val);
        }
        double[] res = new double[queries.size()];
        for (int i = 0; i < queries.size(); i++) {
            String x = queries.get(i).get(0);
            String y = queries.get(i).get(1);
            if (!map.containsKey(x) || !map.containsKey(y)) res[i] = -1.0;
            else res[i] = calculateHelper(x, y, map, new HashSet<String>());
        }
        return res;
    }


    public double calculateHelper(String start, String end, HashMap<String, HashMap<String, Double>> map, HashSet<String> visited) {
        if (start.equals(end)) return 1.0;
        if (!map.containsKey(start)) return -1.0;
        visited.add(start);
        for (String n : map.get(start).keySet()) {
            if (visited.contains(n)) continue;
            double res = calculateHelper(n, end, map, visited);
            if (res > 0) return res * map.get(start).get(n); // 这里也很重要
        }
        return -1.0;
    }
}

File: Leetcode/src/com/leetcode/graph/_399_evaluate_division.py
from typing import List


class Solution:
    # time:O(V + E) space:O(n)
    def calcEquation(
        self, equations: List[List[str]], values: List[float], queries: List[List[str]]
    ) -> List[float]:
        graph = {}
        for i in range(len(values)):
            first, second = equations[i]
            if first not in graph:
                graph[first] = {}
            graph[first][second] = values[i]
            if second not in graph:
                graph[second] = {}
            graph[second][first] = 1.0 / values[i]

        # 这里并不需要backtracking
        def dfs(start, end, visited):
            if start not in graph or end not in graph:
                return -1.0
            if start == end:
                return 1

            visited.add(start)
            for adj in graph[start]:
                if adj not in visited:
                    curr_val = graph[start][adj]
                    next_val = dfs(adj, end, visited)
                    if next_val > 0:
                        return curr_val * next_val
            return -1.0

        res = []
        for q in queries:
            curr = dfs(q[0], q[1], set())
            res.append(curr)
        return res
